lib/internal/Magento/Framework/Cache/LockGuardedCacheLoader.php

```diff
@@ -93,6 +111,62 @@ public function lockedLoadData(
         return $cachedData;
     }
 
+    /**
+     * Loads cached data in non blocking way
+     *
+     * When data is not cached it will try to grab a lock.
+     * If lock is not obtainable it just returns back uncached data,
+     * so connection is not going into cyclic deadlock
+     *
+     * @param string $lockName
+     * @param callable $dataLoader
+     * @param callable $dataCollector
+     * @param callable $dataSaver
+     * @param callable|null $dataFormatter
+     * @return mixed
+     */
+    public function nonBlockingLockedLoadData(
+        string $lockName,
+        callable $dataLoader,
+        callable $dataCollector,
+        callable $dataSaver,
+        callable $dataFormatter = null
+    ) {
+        $cachedData = $dataLoader();
+
+        if ($cachedData !== false) {
+            return $cachedData;
+        }
+
+        $isLocked = $this->isLocked($lockName);
```

Comment:
Is this check necessary or could it simply be skipped? The attempt to lock on line 152 should be enough, or? (especially given that it's non-blocking).

The problem I see is a race-condition between this line check and the line 149, where the $isLocked result is being used. In the time that passes between those two lines, the actual lock state could have changed.

Locks state is usually best derived by actually trying to acquire the lock (on success, you are guaranteed to hold the lock; on failure you have a guarantee that you are not holding the lock). In most cases where an isLocked test is used to determine the state of a lock, there is a race condition between determining the state and being able to evaluate the result of the state-check.

Reply:
This was tested by at least 70 load tests. Is locked is used to check if stale data need to be loaded instead of trying to acquire a lock for writing. If this check is removed then you get more then one process acquiring a lock under high load that can result in cyclic cache writes as was shown in load test results.

Reply:
Basically `isLocked` prevents race condition in this case, as you don't try to acquire the lock if during loading of cache data another process already acquired a lock and on the moment when you reach write phase lock is already released by another process so your process starts to think lock can be acquired again. See test cases for this method, they cover all mentioned use cases uncovered during highly concurrent load tests. If you remove this check - tests will fail.

Reply:
The lock manager should guarantee that that only 1 process can acquire a lock at any time, so all except the one that got the lock, should fail acquiring the lock, right? Those that did not get the lock (essentially meaning the same as isLocked being false) can then use that opportunity to load stale data, or?
i.e.
```php
$isLockAcquired = $this->locker->lock($lockName, 0);
if (!$isLockAcquired) {
    if ($cachedData = $dataLoader()) {
        return $cachedData;
    }
}
//... otherwise do the rest as per current code

Reply:
Just saw your 2nd comment above after writing my comment. I will dig a bit into the load tests to see what is being tested for.

Reply:
@vaimo-wilko that is only true when you acquire a lock on reading as well (that is a bad idea in this case). Lock manager does its job well for write lock acquiring but it does not mean that at the microseconds between cache write and lock release another process didn't start check for cache existence and without finding data in cache starts to acquire a lock, but because the lock was released it successfully does it.